homeassistant/components/reolink/__init__.py

```diff
@@ -100,7 +100,11 @@ async def async_check_firmware_update() -> (
         async with asyncio.timeout(host.api.timeout * (RETRY_ATTEMPTS + 2)):
             try:
                 return await host.api.check_new_firmware()
-            except (ReolinkError, asyncio.exceptions.CancelledError) as err:
+            except (
+                ReolinkError,
+                asyncio.CancelledError,
```

Comment:
What is raising CancelledError? That's not expected in code like this.

Reply:
To be honest: I don't really know.

But I see users that block internet access reporting this:
```
Logger: homeassistant.config_entries
Source: config_entries.py:406
First occurred: 4:38:52 PM (6 occurrences)
Last logged: 5:21:49 PM

Error setting up entry Entry Doorbell for reolink
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/config_entries.py", line 406, in async_setup
    result = await component.async_setup_entry(hass, self)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/components/reolink/__init__.py", line 138, in async_setup_entry
    await asyncio.gather(
asyncio.exceptions.CancelledError
```

It almost certanly has to be comming form the firmware check because that's the only thing that requires internet acces.

Reply:
Now that I am ignoring Exceptions in the firmware check anyway, I could get rid of catching the asyncio.CancelledError altogether. Only the dataupdatecoordinator will not get a nice `UpdateFailed` error message but a exception I think.

Reply:
`asyncio.CancelledError` is a subclass of `BaseException` not `Exception`.

https://docs.python.org/3/library/asyncio-exceptions.html#asyncio.CancelledError

I'll look a bit more.

Reply:
Is that the full stack trace above?

Reply:
I believe so, it comes from this issue: https://github.com/home-assistant/core/issues/110315
The user did not provide anything else.

Reply:
It is related to this old PR: https://github.com/home-assistant/core/pull/91738
That has a issue linked with a bigger stack trace: https://github.com/home-assistant/core/issues/93174
```

Logger: homeassistant.config_entries
Source: components/reolink/__init__.py:121
First occurred: 09:49:46 (1 occurrences)
Last logged: 09:49:46

Error setting up entry BackPorch for reolink
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/update_coordinator.py", line 258, in _async_refresh
    self.data = await self._async_update_data()
  File "/usr/src/homeassistant/homeassistant/helpers/update_coordinator.py", line 217, in _async_update_data
    return await self.update_method()
  File "/usr/src/homeassistant/homeassistant/components/reolink/__init__.py", line 98, in async_check_firmware_update
    return await host.api.check_new_firmware()
  File "/usr/local/lib/python3.10/site-packages/reolink_aio/api.py", line 1532, in check_new_firmware
    json_data = await self.send(body, expected_response_type="json")
  File "/usr/local/lib/python3.10/site-packages/reolink_aio/api.py", line 3262, in send
    return await self.send_chunk(body, param, expected_response_type, retry)
  File "/usr/local/lib/python3.10/site-packages/reolink_aio/api.py", line 3345, in send_chunk
    response = await self._aiohttp_session.post(url=self._url, json=body, params=param, allow_redirects=False)
  File "/usr/local/lib/python3.10/site-packages/aiohttp/client.py", line 560, in _request
    await resp.start(conn)
  File "/usr/local/lib/python3.10/site-packages/aiohttp/client_reqrep.py", line 899, in start
    message, payload = await protocol.read()  # type: ignore[union-attr]
  File "/usr/local/lib/python3.10/site-packages/aiohttp/streams.py", line 616, in read
    await self._waiter
asyncio.exceptions.CancelledError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/config_entries.py", line 387, in async_setup
    result = await component.async_setup_entry(hass, self)
  File "/usr/src/homeassistant/homeassistant/components/reolink/__init__.py", line 121, in async_setup_entry
    await asyncio.gather(
asyncio.exceptions.CancelledError
```

Reply:
Thanks for mentioning `BaseException`, good to know!

Reply:
We were already catching `asyncio.CancelledError` here before #110315. Then it shouldn't leak here if it would be `host.api.check_new_firmware` that is the source of it.

Reply:
But what else could it be?
In any case this PR will ensure no exception or cancellation from the firmware coordinator can stop the setup.

Will the setup of a integration be cancelled by higher level HomeAssistant code if it takes too long?
If so, what is the maximum allowed time?

Reply:
will the `asyncio.gather(..., return_exceptions=True)` change in this PR solve this then?

Reply:
The max time for stage 2 integrations is 300 seconds. There should be a warning logged if that happens.

> "Setup timed out for stage 2 - moving forward"

Reply:
I suggest we remove catching `asyncio.CancelledError` here and the uncancelling below. Since we gather now with returning exceptions for the first refresh we'll catch any exception from that coordinator during config entry setup. If it happens at a later stage, we can investigate that stack trace further.

Reply:
> The max time for stage 2 integrations is 300 seconds. There should be a warning logged if that happens.

`asyncio.timeout(host.api.timeout * (RETRY_ATTEMPTS + 2))` == `asyncio.timeout(150)`
So it can not take longer than 150 seconds, so we should be well clear off the 300 seconds.

Reply:
> I suggest we remove catching asyncio.CancelledError here and the uncancelling below.

Done, thank you very much for looking into this together with me!